src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,316 @@
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { URI } from 'vscode-uri';
+import { Connection } from 'vscode-languageserver';
+import { v4 as uuidv4 } from 'uuid';
+import { asyncExec } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+import { IContianerEngine } from '../interfaces/extensionSettings';
+
+
+export class ExecutionEnvironment {
+    private connection: Connection;
+    private context: WorkspaceFolderContext;
+    private _container_engine: IContianerEngine;
+    private _container_image: string;
+    private _container_image_id: string;
+
+    constructor(connection: Connection, context: WorkspaceFolderContext) {
+      this.connection = connection;
+      this.context = context;
+    }
+
+    public async initialize(): Promise<void> {
+      try {
+        const settings = await this.context.documentSettings.get(
+          this.context.workspaceFolder.uri
+        );
+        this._container_image = settings.executionEnvironment.image
+        this._container_engine = settings.executionEnvironment.containerEngine
+        if (this._container_engine === 'auto') {
+          for (const ce of ['podman', 'docker']) {
+              try {
+                child_process.execSync(`which ${ce}`, {
+                  encoding: 'utf-8'
+                })
+                this._container_engine = <IContianerEngine>ce;
+                this.connection.console.log(`Container engine set to: '${ce}'`)
+                break
+              } catch (error) {
+                this.connection.console.error(`Container engine '${ce}' not found`)
+              }
+            }
+        } else {
+          try {
+            child_process.execSync(`which ${this._container_engine}`, {
+              encoding: 'utf-8'
+            })
+          } catch (error) {
+            this.connection.window.showErrorMessage(`Container engine '${this._container_engine}' not found. Failed with error '${error}'`)
+            return
+          }
+        }
+        if (!['podman', 'docker'].includes(this._container_engine)) {
+          this.connection.window.showInformationMessage(`Valid contianer engine not found ${this._container_engine}.`)
+          return
+        }
+        const imagePuller = new ImagePuller(
+          this.connection,
+          this._container_engine,
+          this._container_image,
+          settings.executionEnvironment.pullPolicy
+        )
+        const setupDone = await imagePuller.setupImage()
+        if (!setupDone) {
+          this.connection.window.showWarningMessage(`Execution environment image ${this._container_image} setup failed.`)
```

Comment:
So we'll have a warning window (from here) and an error window (from setupImage()). That's too much.

Reply:
Could we do the other way around, so that we have window interactions in the service, instead of the helper? Essentially, we've already seen one or two instances of double-logging/messaging and if we spread that capability/responsibility around the code, it will be much harder to avoid.

Reply:
Make sense. Fixed it in latest commit.